packages/button/src/Button.tsx

```diff
@@ -289,128 +288,94 @@ const disabledStyle = css`
-  href: string;
-}
-
-interface ButtonButtonProps
-  extends HTMLElementProps<'button'>,
-    SharedButtonProps {
-  href?: null;
-}
-
-type CustomElementButtonProps = SharedButtonProps & {
-  as: React.ElementType<any>;
-  [key: string]: any;
-};
-
-type ButtonProps =
-  | LinkButtonProps
-  | ButtonButtonProps
-  | CustomElementButtonProps;
-
-function usesCustomElement(
-  props: ButtonProps,
-): props is CustomElementButtonProps {
-  return (props as any).as != null;
-}
-
-function usesLinkElement(
-  props: LinkButtonProps | ButtonButtonProps,
-): props is LinkButtonProps {
-  return props.href != null;
-}
-
-const Button = React.forwardRef((props: ButtonProps, forwardRef) => {
-  const { usingKeyboard: showFocus } = useUsingKeyboardContext();
-
-  const {
-    className = '',
-    children = null,
-    disabled = false,
-    variant = Variant.Default,
-    size = Size.Normal,
-    glyph,
-  } = props;
-
-  const commonProps = {
-    className: cx(
-      baseStyle,
-      buttonSizes[size],
-      buttonVariants[variant],
-      { [disabledStyle]: disabled },
-      { [focusStyle]: showFocus },
+const Button: ExtendableBox<BaseButtonProps> = React.forwardRef(
+  (
+    {
       className,
-    ),
-    // only add a disabled prop if not an anchor
-    ...(!usesLinkElement(props) && { disabled }),
-    'aria-disabled': disabled,
-  };
-
-  const rest = omit(props, [
-    'as',
-    'className',
-    'disabled',
-    'size',
-    'variant',
-    'children',
-    'glyph',
-  ]);
-
-  const spanStyle = css`
-    // Usually for this to take effect, you would need the element to be
-    /* ￿positioned￿. Due to an obscure part of CSS spec, flex children */
-    // respect z-index without the position property being set.
-    //
-    // https://www.w3.org/TR/css-flexbox-1/#painting
-    z-index: 1;
-    display: inline-flex;
-    flex-grow: 1;
-    justify-content: center;
-    align-items: center;
-  `;
-
-  const modifiedGlyph =
-    glyph && children
-      ? React.cloneElement(glyph, {
-          className: cx({ [glyphMargins[size]]: glyph != null }),
-        })
-      : glyph;
-
-  const renderButton = (Root: React.ElementType<any> = 'button') => (
-    <Root
-      ref={forwardRef}
-      type={Root === 'button' ? 'button' : undefined}
-      {...(rest as HTMLElementProps<any>)}
-      {...commonProps}
-    >
+      children,
+      disabled = false,
+      variant = Variant.Default,
+      size = Size.Normal,
+      glyph,
+      ...rest
+    }: BaseButtonProps,
+    ref: React.Ref<any>,
+  ) => {
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+
+    const commonProps = {
+      ref,
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        { [disabledStyle]: disabled },
+        { [focusStyle]: showFocus },
+        className,
+      ),
+      // only add a disabled prop if not an anchor
+      ...(rest.href === undefined && { disabled }),
```

Comment:
Thoughts on this vs `rest.href == null` to cover both cases?

Reply:
WFM 👍